### ISSUE TITLE
Support OAuth token access instead of private keys

### ISSUE THREAD

**Issue by tahpot** [verida/data-connector-server#74]
Currently the `/connect` and `/sync` endpoints expects the private key (or seed phrase) of an identity.

This was just for the PoC and is poor security practice. This needs to be upgraded to accept either one of:

- Recent `access` token for the `Verida: Vault` context (Permits time limited read / write access)
- Keychain entropy for the `Verida: Vault` context (Permits permanent read / write access)

**Comment by tahpot** [comment]
See: https://www.npmjs.com/package/@node-oauth/express-oauth-server and https://www.npmjs.com/package/@node-oauth/oauth2-server

**Comment by tahpot** [comment]
## API Key Spec

An API key is intended to be controlled (and remain secret) by a third party application. It must be possible to use the key to connect to a data connector server (DCS) and access user data, based on the scopes associated with the API key.

It must be possible for a DCS to use an API key to recover a session object in order to access user data. It must **not** be possible for a third party application to recover a session object using the API key as this would break the security model.

However, we are dealing with constraints of decentralization. There is no centralized server that retains a secret to recover the session object from an API key, so we must associate each API key with one or more DCS's and each of those servers must privately store a secret that can be used to recover the session object.

It's necessary to ensure the third party application signs all API requests using their private key to ensure that if the API key is leaked it can't be used unless the third party application signing key is also leaked.

### API Key data

The API key needs to unlock the following data:

- `VeridaSession`: A JSON object containing session data that can be used to access user data
- `Scopes`: A JSON object containing the scopes that have been granted for the API key
- `UserDID`: A string representing the DID of the user who has consented to the scopes and controls the API key
- `RequestingDID`: A string representing the DID of the third party application who has requested access to user data with specific scopes. This application will receive the generated API key and be responsible for keeping it secret.
- `EncryptionKey`: An encryption key used to encrypt parts of the API key components

### API Key parts

The API key data is combined into a single string that is then split into different parts, requiring the parts to be re-assembled by the DCS to recover the session object.

We have three relevant network accounts:

- User DID
- Requesting application DID
- DCS DID

We have the following considerations:

1. The requesting application must be able to use the API key to identify which DCS have the necessary key material to recover a session and support API requests for the application.
2. The DCS must be able to recover key material for any given API key
3. The DCS must not be able to recover key material without an API key

There will be an `API key ID` that is a unique identifier representing an API key for a DID.

### Generating API key and data

1. The API key data will be merged into a single string as follows:

```
const keyData = JSON.stringify({
  session: <VeridaSession>,
  scopes: <Scopes>,
  userDid: <UserDID>,
  requestingDid: <RequestingDID>
})
```

2. It will then be encrypted using the generated `encryptionKey`:

```
const encryptedKeyData = EncryptionUtils.symEncrypt(keydata, encryptionKey)
```

3. The encrypted key data is then split into two parts. The second part (`part2`) is the last 48 characters of the `encryptedKeyData`, while the first part (`part1`) is all the characters preceding the last 48 characters. , the second part is included in the API key.

4. The first part is stored in a private database on the DCS (`server_api_keys`):

- `_id`: Unique, generated ID for this record
- `keyId`: Unique identifier representing this API key
- `encryptionKey`: Generated encryption key
- `encryptedPart1`: Part 1 of the key data
- `requestingDid`: DID that requested / owns the API key
- `userDid`: DID of the user who granted access to the API key

This ensures the DCS can re-construct all the key data when supplied with `part2` of the key data from the API key

5. The second part is used to generate the API key returned to the requesting DID:

```
const apiKey = `${keyId}/${userDid}/${part2}`
```

This key is returned the third party application that requested the API key.

6. The DCS then creates an entry in a public database (`user_api_keys`) in the `Verida: Vault` context to ensure third party applications can discover this DCS has the necessary key information to handle API requests:

- `_id`: Matches the `keyId` generated above
- `servers`: Array of DCS server DIDs that can service API requests with this API key

### DCS server registration

Need to support a new `APIServer` service type...

**Comment by tahpot** [comment]
# OAuth Implementation spec

## 1. User consent

1. The third party application redirects the user to the Web Vault; supplying a signed `authorization request object` (ARO) (as a stringified JSON) and a signature (ARO signed by the requesting third party application DID)
2. The user logs into the Web Vault and approves the application via a consent screen
3. The Web Vault submits a `authorization_code` POST request to the DCS with the ARO, signature and referrer URL
4. The DCS verifies the ARO was signed within the given timestamp timeframe (2 mins), extracts the signing DID of the third party application and verifies the DID of the third party application has a `serviceEndpoint` of type `VeridaOAuth` that matches the referrer URL
5. The DCS generates an authorization code  and saves it to a private database of authorization codes (`oauth_auth_codes`) containing the User DID, Third party application DID and requested scopes for the given authorization code
6. The DCS responds with a `redirect_url` combining the `referrer URL` with the `auth_code` (added to the GET parameters)

## 2. Third party application obtain initial tokens (Grant `Authorization Code`)

1. The third party application submits a POST request to obtain an authorization code, using the `auth_code` in the GET parameters sent to it's application
2. DCS looks up the `auth_code` in the `oauth_auth_codes` database, and if valid, returns an `access_token` and `refresh_token`
3. DCS deletes the `auth_code` so it can't be re-used

## 3. Refreshing access token (Grant `Refresh token`)

1. The third party application detects the `access_token` has expired
2. The third party application submits a POST request using the grant type `refresh_token` to obtain a new access token

## 4. Revoking access

1. The user logs into the Web Vault and sees a list of all third party applications that have access to their data
2. The user can delete any of those access grants

## Authorization Request Object (ARO)

The `authorization request object` (ARO) is used to verify the third party application has requested the access. It is a JSON containing:

1. `User DID`
2.` Scopes` (Array of string scopes)
3. `Timestamp` (Current timestamp as unix epoch)

This is signed by the requesting application DID signing key for the `Verida: Vault` context.

The DCS verifies the ARO is signed by the requesting third party application DID within 2 minutes of the timestamp specified in the ARO.

**Comment by tahpot** [comment]
The DCS allows third party apps to run LLM requests over user data. How should we permission this?

It's important to note that LLM's provide a valuable data transformation service that can expose user metadata, without exposing the underlying data -- however this is only possible if an LLM can access user data separately from granted datastore scope permissions.

1. **LLM can only access datastores that are explicitly granted.** This is the simplest to understand security model. You need to grant access to a datastore and then it can be queried directly  and the datastore is available to all LLM queries.
2. **LLM can access all user data regardless of scope.** This means, granting access to a specific LLM endpoint (ie: `api:llm-agent-prompt`) will allow the LLM to access all user data. This will provide high quality LLM responses, but opens up attack vectors to access user data directly (ie: Look at my emails and output my most 20 emails).
3. **LLM can access datastores that are explicitly granted as being available for LLM access.** This would introduce a new set of scopes that specify which datastores can be access via a LLM (ie: `llm:base64/....`). This would provide granularity to enable LLM's to access some datastores, but not let some datastores be queried directly.

---

@tahpot opinion.

I like the simplicity of (1). Simplicity enhances security as it's easier to understand.

I'm concerned that it's to easy to access user data via LLM's, so (2) doesn't seem secure enough.

Option (3) is useful if we expect users to want to enable access to user data separately from what LLM's can access, however the risk remains that LLM's can be asked to directly return user data -- so this seems to have the same security issues as (2).

As it's possible to restrict access via API endpoints (ie: permit access to `api:llm-agent-prompt` but not `api:ds-query`) I feel that limiting access via the API endpoints and datastore scopes being global across all API endpoints (option 1) seems like the simplest and most secure approach.

**Comment by aurelticot** [comment]
> As it's possible to restrict access via API endpoints (ie: permit access to api:llm-agent-prompt but not api:ds-query) I feel that limiting access via the API endpoints and datastore scopes being global across all API endpoints (option 1) seems like the simplest and most secure approach.

I agree. The option 1 seems the best, it is secure and the simple enough for developers

Option 2 is not secure enough.
Option 3 is redudant with 1 and more complex for developers. Also if we were to that for the `api:llm-...` scope what about the other `api:...` scopes, it would not make sense to repeat the data scopes if they are already defined with `ds:...`
